src/main/scala/rules/Executor.scala

```diff
@@ -91,69 +99,113 @@ object executor extends ExecutionRules {
   private def follows(s: State,
                       edges: Seq[SilverEdge],
                       @unused pvef: ast.Exp => PartialVerificationError,
-                      v: Verifier)
+                      v: Verifier,
+                      joinPoint: Option[SilverBlock])
                      (Q: (State, Verifier) => VerificationResult)
                      : VerificationResult = {
 
-    if (edges.isEmpty) {
-      Q(s, v)
-    } else if (edges.length == 1) {
-      follow(s, edges.head, v)(Q)
-    } else {
-      edges match {
-        case Seq(thenEdge@ConditionalEdge(cond1, _, _, _), elseEdge@ConditionalEdge(cond2, _, _, _))
-            if Verifier.config.parallelizeBranches() && cond2 == ast.Not(cond1)()  =>
-          val condEdgeRecord = new ConditionalEdgeRecord(thenEdge.condition, s, v.decider.pcs)
-          val sepIdentifier = v.symbExLog.openScope(condEdgeRecord)
-          val res = eval(s, thenEdge.condition, IfFailed(thenEdge.condition), v)((s2, tCond, v1) =>
-            brancher.branch(s2, tCond, Some(thenEdge.condition), v1)(
-              (s3, v3) => {
-                val s3p = handleOutEdge(s3, thenEdge, v3)
-                exec(s3p, thenEdge.target, thenEdge.kind, v3)((s4, v4) => {
-                  v4.symbExLog.closeScope(sepIdentifier)
-                  val branchRes = Q(s4, v4)
-                  branchRes
-                })
-              },
-              (s3, v3) => {
-                val s3p = handleOutEdge(s3, elseEdge, v3)
-                exec(s3p, elseEdge.target, elseEdge.kind, v3)((s4, v4) => {
-                  v4.symbExLog.closeScope(sepIdentifier)
-                  Q(s4, v4)
-                })
-              }))
-          res
-        case _ =>
-          val uidBranchPoint = v.symbExLog.insertBranchPoint(edges.length)
-          val res = edges.zipWithIndex.foldLeft(Success(): VerificationResult) {
-            case (result: VerificationResult, (edge, edgeIndex)) =>
-              if (edgeIndex != 0) {
-                v.symbExLog.switchToNextBranch(uidBranchPoint)
-              }
-              v.symbExLog.markReachable(uidBranchPoint)
-              result combine follow(s, edge, v)(Q)
+    // Find join point if it exists.
+    val jp: Option[SilverBlock] = edges.headOption.flatMap(edge => s.methodCfg.joinPoints.get(edge.source))
+
+    (edges, jp) match {
+      case (Seq(), _) => Q(s, v)
+      case (Seq(edge), _) => follow(s, edge, v, joinPoint)(Q)
+      case (Seq(edge1, edge2), Some(newJoinPoint)) if
+          Verifier.config.moreJoins() &&
+          // Can't directly match type because of type erasure ...
+          edge1.isInstanceOf[ConditionalEdge[ast.Stmt, ast.Exp]] &&
+          edge2.isInstanceOf[ConditionalEdge[ast.Stmt, ast.Exp]] &&
```

Comment:
Doesn't `isInstanceOf` suffer from all the same problems as match expressions regarding erasure? In other words, why does `isInstanceOf` here work?

I wonder if we would need sth stronger like `TypeTag` to preserve the erased types at runtime (https://www.baeldung.com/scala/type-information-at-runtime)

Reply:
I think it has that problem in principle, yes, but I'm 99% sure it's not an issue here because we just need to know it's a ``ConditionalEdge``; the type arguments are always the same, so if it's a ``ConditionalEdge``, it's definitely a ``ConditionalEdge[ast.Stmt, ast.Exp]``.